File: src/main/java/com/pygostylia/osprey/Explosion.java
package com.pygostylia.osprey;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Random;

public class Explosion {
    static Random rng = new Random();

    static Collection<Location> generateBoomBlocks(Location center, float power) {
        var locations = new ArrayList<Location>();
        final int maxRadius = 20;
        for (int x = -maxRadius; x < maxRadius; x++) {
            for (int y = -maxRadius; y < maxRadius; y++) {
                for (int z = -maxRadius; z < maxRadius; z++) {
                    var location = center.offset(x, y, z);
                    if (location.y() > 255 || location.y() < 0) continue;
                    double distance = center.distance(location);
                    if (power / 1.5 > distance || rng.nextDouble() * power > distance) {
                        locations.add(location);
                    }
                }
            }
        }
        return locations;
    }
}

File: src/main/scala/com/pygostylia/osprey/Explosion.scala
package com.pygostylia.osprey

import java.util
import java.util.Random
import scala.jdk.CollectionConverters.IterableHasAsJava

object Explosion {
  private[osprey] val rng = new Random

  def generateBoomBlocks(center: BlockPosition, power: Float, maxRadius: Int = 20): util.Collection[BlockPosition] = {
    var points = for (x <- -maxRadius to maxRadius;
                      y <- -maxRadius to maxRadius;
                      z <- -maxRadius to maxRadius) yield center.offset(x, y, z)
    points = points.filter(p => 0 to 255 contains p.y)
    points = points.filter({ p =>
      val distance = center.distance(p)
      power / 1.5 > distance || rng.nextDouble() * power > distance
    })
    points
  }.asJavaCollection
}
